### PR TITLE
Add Firewall Settings UI

### PR DESCRIPTION
**Pull request by nateweller**
This PR adds a new `firewall` module to Jetpack, and related settings UI. This also adds a check to the WAF `run.php` file to only run when the module is enabled.

#### Changes proposed in this Pull Request:

* Adds a new `firewall` Jetpack module
* Adds a new class to the WAF package, that registers a new `jetpack/v4/waf` endpoint which provides WAF settings data from the server (`bootstrapPath` and `hasRulesAccess`).
* Adds `<Firewall />` settings component that outputs all of the new settings UI.
* Updates the WAF package `rules.php` to abort execution if the `firewall` module is disabled, and adds a method for providing the path to `bootstrap.php`.

#### Jetpack product discussion

1201069996155217-as-1202025817736945/f

p1HpG7-ffB-p2

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Boot up a Jetpack site running this branch and visit `/wp-admin/admin.php?page=jetpack#/settings`

**UI**
* Validate UI against the designs found in p1HpG7-ffB-p2.

**Module Toggling**
* Validate that the module is enabled by default.
* Generate the `bootstrap.php` from the CLI and validate that the firewall runs by default.
* Disable the module and validate that the firewall does not run.
* Validate that all setting toggles and text boxes save and load correctly.

**Upgrade Banner**
* Validate that the banner is not displayed on a site containing a plan with Jetpack Scan.
* Validate that the banner is displayed on a site not containing a plan with Jetpack Scan.
* Validate that the upgrade button takes you to the Scan/Security upgrade page.

**Enhance Protection Info**
* Validate that the path to `bootstrap.php` is correct.

#### Screenshots:

Default with no Scan plan:

Screen Shot 2022-04-14 at 10 33 50 AM

All settings enabled with Scan plan:

Screen Shot 2022-04-14 at 10 33 04 AM

All settings disabled:

Screen Shot 2022-04-14 at 10 31 55 AM

### CODE REVIEW

File: projects/plugins/jetpack/modules/firewall.php
Comment (nateweller) [inline projects/plugins/jetpack/modules/firewall.php:5]: Any framework of rhyme or reason I should be considering when picking this number?

File: projects/plugins/jetpack/modules/firewall.php
Comment (nateweller) [inline projects/plugins/jetpack/modules/firewall.php:3]: Do we want to be naming things as "WAF" and not "Firewall" everywhere in the codebase? I chose "Firewall" here as that is the actual title of the module that the user interacts with.

**Comment by nateweller** [comment]
("All commits were linted before commit" is just due to reverts)

**Comment by singerb** [comment]
I'm trying to test this but running into issues. However, so far the PHP side of this looks good. @miguelxpn or @bhoop can you take a look at the React side?

File: projects/plugins/jetpack/_inc/client/security/firewall.jsx
Comment (miguelxpn) [inline projects/plugins/jetpack/_inc/client/security/firewall.jsx:46]: Would you mind changing the option names to `jetpack_waf_ip_allow_list` and `jetpack_waf_ip_block_list`? After talking with @ArSn while looking at the backend site he reminded that we already have some options that start with jetpack_waf, so having some that starts with jetpack_firewall would be a bit odd. Sorry about asking this after agreeing with the naming. 😅

**Review by miguelxpn — CHANGES_REQUESTED** [review]

File: projects/plugins/jetpack/modules/firewall.php
Comment (ArSn) [reply projects/plugins/jetpack/modules/firewall.php:3]: Wouldn't it be easier if we called it "WAF" everywhere? What would be the distinction between when we call it WAF and when we call it firewall? At least on the technical side, it has been WAF everywhere that I am aware of so far.

File: projects/plugins/jetpack/_inc/client/security/firewall.jsx
Comment (ArSn) [inline projects/plugins/jetpack/_inc/client/security/firewall.jsx:133]: as of now, the IP checks do not support ranges. We are expecting an array of single IP addresses, so it would probably make sense to reflect that somehow here (e.g. either 1 IP per line or comma separated). Ranges and CIDR blocks will certainly come in a future version.

File: projects/plugins/jetpack/_inc/client/security/firewall.jsx
Comment (ArSn) [inline projects/plugins/jetpack/_inc/client/security/firewall.jsx:151]: same thing here about the IP ranges

**Comment by ArSn** [comment]
Merged master into this branch to have the IP block list and allow list server side functionality in this branch in order to base the feature-interconnectivity work on this branch while the PR is still in review. See 1201069996155224-as-1202148145416678

**Comment by ArSn** [comment]
After clarifying the changes mentioned in my last comment further - see p1650458364845089-slack-CBG1CP4EN and p1650460731138599-slack-CBG1CP4EN - I actually like the solution to ignore the sanitation check better since the sanitation does not make much sense to begin with in this particular instance. I merged the master in here again, resolving the conflict and having the linting now safe on all sides. Sorry again for the unintended side effects of my original merge 😬

File: projects/plugins/jetpack/_inc/client/security/firewall.jsx
Comment (nateweller) [reply projects/plugins/jetpack/_inc/client/security/firewall.jsx]: Updated!

File: projects/plugins/jetpack/_inc/client/security/firewall.jsx
Comment (nateweller) [reply projects/plugins/jetpack/_inc/client/security/firewall.jsx]: I've updated the placeholder here to display a list of single IP addresses :+1:

File: projects/plugins/jetpack/_inc/client/security/firewall.jsx
Comment (nateweller) [reply projects/plugins/jetpack/_inc/client/security/firewall.jsx]: I've updated the placeholder here to display a list of single IP addresses :+1:

**Comment by nateweller** [comment]
Addressed the following:
* Standardized naming to use "waf" instead of "firewall" in the code, while still presenting to user as "Firewall". 
* Updated IP list placeholders.

**Review by ArSn — DISMISSED** [review]
I can really only speak to the PHP part, but this one LGTM as well. Once the review from @miguelxpn is in, I think this is good to merge.

**Review by miguelxpn — DISMISSED** [review]
The frontend stuff looks good to me! The only thing we need to figure out is the whole module activation issue.

**Comment by ArSn** [comment]
Merged master again to get the PR check green.

File: projects/packages/waf/src/class-waf-runner.php
Comment (dereksmart) [inline projects/packages/waf/src/class-waf-runner.php:75]: This `Jetpack` plugin method is only a wrapper for the `Modules()->is_active()` [package method](https://github.com/Automattic/jetpack/blob/master/projects/packages/status/src/class-modules.php#L25). Does it make sense to use that now to avoid refactoring when we bring it into the Protect plugin (assuming we are)?

Comment (nateweller) [reply projects/packages/waf/src/class-waf-runner.php]: Removed the usage of the `Jetpack` method via https://github.com/Automattic/jetpack/pull/23769/commits/5b4c7a07b1970d2da98c4b0d61ba047b60a3d717 :+1:

**Review by ArSn — DISMISSED** [review]
Still LGTM.

**Comment by kraftbj** [comment]
@nateweller any other changes pending for this one?

**Comment by nateweller** [comment]
Based on p1650556975439039-slack-CBG1CP4EN, I've moved the endpoints into the WAF package. This will not only be useful when we need these for the Protect plugin, but also fixes an issue where the UI would not be displayed when the module was disabled, due to the `uiEnabled` value coming from the module. Now that it's in the package, this will not be a problem.

@kraftbj This is the last of the changes that I have planned!

**Comment by kraftbj** [comment]
@nateweller Sounds good. Let me give it a last check and I'll approve it right away.

**Review by dereksmart — CHANGES_REQUESTED** [review]
Gave it a quick spin and noticed a some behaviors that don't look intended: 
1. When the module is _inactive_, it does not show in the main settings at all. 
2. When the `JETPACK_WAF_UI` is _not_ set, you can still see and "activate" it from two places: 
The first is in `/wp-admin/admin.php?page=jetpack_modules`
The other is in wp cli, via `wp jetpack module list`, and you can activate it via `wp jetpack module activate waf`

**Comment by nateweller** [comment]
@dereksmart Thanks for looking at this! 

#1 should be addressed in https://github.com/Automattic/jetpack/pull/23769/commits/9ba215cc00c8a23dcc81800322063877f38c2b54, which I think I may have pushed while you were reviewing? 

I've prompted some discussion with the team for #2, it's been mentioned that we may potentially remove the define sooner than later which would negate this one. Otherwise, I will have to look into how we could conditionally make the module available based on the constant.

**Comment by kraftbj** [comment]
The blocker I see:

Without the UI flag, the module is shown as enabled already in the modules list and is actually enabled (e.g. I'm seeing `Closure on line 39 of wp-content/plugins/jetpack-dev/jetpack_vendor/automattic/jetpack-waf/actions.php` being fired on `plugin_loaded`.

To merge this, we should have it off until the gate is removed and/or have the gate also short-circuit the module itself if it is not activate (related to Derek's point).

**Comment by nateweller** [comment]
Per some discussion (p1650575200897379-slack-C029WFNV69M) and our release plan (pbuNQi-2Oa-p2), I've gone ahead and removed the `JETPACK_WAF_UI` gate completely as we are now planning to include the UI in the public beta.

File: projects/plugins/jetpack/modules/waf.php
Comment (kraftbj) [inline projects/plugins/jetpack/modules/waf.php:8]: Per the release plan, shouldn't this be No? pbuNQi-2Oa-p2

```suggestion
 * Auto Activate: No
```

File: projects/plugins/jetpack/modules/waf.php
Comment (kraftbj) [inline projects/plugins/jetpack/modules/waf.php:7]: This requires a connection since we get the rules from the API.

```suggestion
 * Requires Connection: Yes
```

**Review by kraftbj — COMMENTED** [review]
I need to go through this with a bit finer-tooth comb and near the end of my day. Very immediately, I think we need to declare this as a connected module and not auto-activated since the intent is this to go beta for 10.9.

**Review by miguelxpn — COMMENTED** [review]
Tested it out and it works great! Now we can see the settings even with the module disabled. Once we address the things @kraftbj pointed out we should be good to go!

File: projects/plugins/jetpack/_inc/client/security/waf.jsx
Comment (ArSn) [inline projects/plugins/jetpack/_inc/client/security/waf.jsx:230]: I just read that in the interface and was like "What? If I don't undo this when the firewall is turned off, I will uninstall Jetpack?" But what it actually means is that I should undo the action in either of those two cases, which does make sense :)

Therefore, small improvement suggestion.
```suggestion
								"Don't forget to undo this action when Firewall is turned off, or when you uninstall Jetpack.",
```

**Review by ArSn — APPROVED** [review]
Only one mini nitpick in addition to what has already been said by the others.

File: projects/plugins/jetpack/_inc/client/security/waf.jsx
Comment (kraftbj) [reply projects/plugins/jetpack/_inc/client/security/waf.jsx:230]: That's a good point. Do you think we should have more proactive notices if someone deactivates the feature while leaving that in place or uninstall Jetpack?

admin notice if disabled with it still active?
e-mail to site admin if still active when Jetpack is deactivated/uninstalled?

**Comment by singerb** [comment]
As the tags indicate, based on pbuNQi-2Oa-p2#comment-4568 don't merge this until after the 10.9 freeze is lifted.

File: projects/plugins/jetpack/_inc/client/security/waf.jsx
Comment (ArSn) [reply projects/plugins/jetpack/_inc/client/security/waf.jsx]: @kraftbj Actually, for the case when the firewall is turned off, we can be smart enough about it that the firewall does not actually do anything (this is not yet happening but should be easy to do, see p1650568832851519/1650548147.844039-slack-C029WFNV69M). For the other case I am not sure how exactly the deactivation/uninstall routine looks. If we do indeed remove all files, making this more prominent sounds like a good idea. Or at least create an empty file that does nothing in its place, if possible.

File: projects/packages/waf/src/class-waf-endpoints.php
Comment (jeherve) [inline projects/packages/waf/src/class-waf-endpoints.php:79]: Should we limit this to site admins instead? Non-admins can currently access the Jetpack dashboard to manage other settings, but I don't think non-admins need to see or update WAF settings. Maybe `jetpack_manage_modules` would be more appropriate here?

File: projects/plugins/jetpack/_inc/client/security/waf.jsx
Comment (nateweller) [reply projects/plugins/jetpack/_inc/client/security/waf.jsx]: Thanks for your review Jeremy! This one has been addressed via https://github.com/Automattic/jetpack/pull/23769/commits/b9a426dee6c272373aff204e196e503791b85309.

File: projects/plugins/jetpack/_inc/client/security/waf.jsx
Comment (nateweller) [reply projects/plugins/jetpack/_inc/client/security/waf.jsx:152]: Addressed via https://github.com/Automattic/jetpack/pull/23769/commits/67a2322e661e4463314cde4b0be25198ea2a5bd8 - used `sprintf` for the IP list with a note for translators, and merged the above privacy strings into a single string with the existing `createInterpolateElement` implementation.

**Review by ArSn — DISMISSED** [review]

**Review by ArSn — APPROVED** [review]

**Review by kraftbj — APPROVED** [review]
Let's give this a go.

File: projects/packages/waf/src/class-waf-endpoints.php
Comment (singerb) [reply projects/packages/waf/src/class-waf-endpoints.php:34]: I think it's only used in the UI, so we'll be fine regardless. We're not trying to gate rules at runtime based on purchases, just in the API and giving people helpful notices in the UI.

**Pull request merged**
